---
title: "PIC"
author: "Hanon McShea"
date: "2024-02-28"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(ape)
library(MASS)
library(nlme)
library(lme4)
require(lmerTest)
library(optimx)
library(ggplot2)
library(MuMIn)
library(plotly)
library(plyr)
library(phytools)
library(scatterplot3d)
library(rgl)
library(gridExtra)
library(caroline)
```

## Read in data files
```{r data}
setwd("C:/Users/hanon/Documents/GitHub/aa_flux/CAIS/")

#read in newick tree
tree<-read.tree(file = "PhylogeneticTree_AllSpecies.nwk.txt")

#read in amino acid frequencies, both proteome or Pfam
pf_aaf<-read.csv(file = "AllPfam_AAC.csv", header = T)
pr_aaf<-read.csv(file = "VertebrateAAC_CAIS.csv", header = T)
pr_aaf<-pr_aaf[,2:24]
pr_aaf<-distinct(pr_aaf, SpeciesUID, .keep_all=T)
cais<-read.csv(file="CAIS_KLD.txt", header=T)
species<-read.csv(file="SpeciesList_11_7_2019.txt",header=T)

colnames(cais)[1] <- 'SpeciesUID'

#decide from here on out whether to use Pfam or proteome
aaf <- pr_aaf

aaf2 <- merge(aaf, species, by.y="SpeciesUID")
aaf3 <- merge(aaf2, cais, by.y="SpeciesUID")
aaf4 <- subset(aaf3, NewickName != "Pantholops_hodgsonii") #remove antelope
```

## Get data in same order as tree tips
```{r PIC prep}
namevector<-as.vector(aaf4$NewickName)#vector of all the vertebrate species names
trimmedtree<-keep.tip(tree,namevector)# keeping only the tips with vertebrate species names
phyloorder <- as.list(trimmedtree$tip.label)
#aafo <- arrange(aaf4, match(NewickName, phyloorder))

aafo <- aaf4[match(phyloorder, aaf4$NewickName),]

aaf4
aafo
```

## Phylogenetically Independent Constrasts
```{r PIC}
#Set up data into vectors for use in PIC
#create named vectors
Avector <- nv(as.vector(aafo$A), aafo$NewickName)
Cvector <- nv(as.vector(aafo$C), aafo$NewickName)
Dvector <- nv(as.vector(aafo$D), aafo$NewickName)
Evector <- nv(as.vector(aafo$E), aafo$NewickName)
Fvector <- nv(as.vector(aafo$"F"), aafo$NewickName)
Gvector <- nv(as.vector(aafo$G), aafo$NewickName)
Hvector <- nv(as.vector(aafo$H), aafo$NewickName)
Ivector <- nv(as.vector(aafo$I), aafo$NewickName)
Kvector <- nv(as.vector(aafo$K), aafo$NewickName)
Lvector <- nv(as.vector(aafo$L), aafo$NewickName)
Mvector <- nv(as.vector(aafo$M), aafo$NewickName)
Nvector <- nv(as.vector(aafo$N), aafo$NewickName)
Pvector <- nv(as.vector(aafo$P), aafo$NewickName)
Qvector <- nv(as.vector(aafo$Q), aafo$NewickName)
Rvector <- nv(as.vector(aafo$R), aafo$NewickName)
Svector <- nv(as.vector(aafo$S), aafo$NewickName)
Tvector <- nv(as.vector(aafo$"T"), aafo$NewickName)
Vvector <- nv(as.vector(aafo$V), aafo$NewickName)
Wvector <- nv(as.vector(aafo$W), aafo$NewickName)
Yvector <- nv(as.vector(aafo$Y), aafo$NewickName)
caisvector<-nv(as.vector(aafo$CAIS), aafo$NewickName)

vectorlist <- list(Avector, Cvector, Dvector, Evector, Fvector, Gvector, Hvector, Ivector, Kvector, Lvector, Mvector, Nvector, Pvector, Qvector, Rvector, Svector, Tvector, Vvector, Wvector, Yvector, caisvector)

vectorlist2 <- list(log(Avector), log(Cvector), log(Dvector), log(Evector), log(Fvector), log(Gvector), log(Hvector), log(Ivector), log(Kvector), log(Lvector), log(Mvector), log(Nvector), log(Pvector), log(Qvector), log(Rvector), log(Svector), log(Tvector), log(Vvector), log(Wvector), log(Yvector), caisvector)

#aafo_pic <- aafo[,-1]
pictable <- data.frame(matrix(nrow = 116))

for (i in vectorlist2) {
  vals <- as.data.frame(pic(i, trimmedtree))
  #colnames(vals) <- gsub('vector', '', deparse(substitute(i)))
  pictable <- cbind(pictable, vals) }

pictable <- pictable[,2:22]
names(pictable) <- c("A", "C", "D", "E", "F", "G", "H", "I", "K", "L", "M", "N", "P", "Q", "R", "S", "T", "V", "W", "Y", "cais")
pictable



######################### pic comparisons, note that the -1 forces the comparisons through the origin

lmiPIC<-lm(pictable$K~ pictable$cais)
summary(lmiPIC)
cor.test(pictable$K, pictable$cais, method = 'pearson')
cor.test(pictable$K, pictable$cais, method = 'spearman')

# Intialize dataframe
AAstart <- which(colnames(pictable) == 'A')
AAend <- which(colnames(pictable) == 'Y')
model_estimates <- data.frame(matrix(nrow = 2))

# loop model over all amino acids
for (aa in AAstart:AAend) {

model <- lm(pictable[,aa]~pictable$cais-1)
model_summary <- summary(model)
model_summary <- as.data.frame(model_summary$coefficients[,1:2])
colnames(model_summary) <- colnames(pictable)[aa]
model_estimates <- cbind(model_estimates, model_summary) }

model_estimates <- model_estimates[,-1]
rownames(model_estimates) <- c('Estimates','Std.Error')
model_estimates <- as.data.frame(t(model_estimates))

model_estimates

#save values, be sure name matches amino acid frequencies you used
#write.csv(model_estimates, 'PICcorrected_Pfam_cais_slopes.csv')
write.csv(model_estimates, 'logPICcorrected_proteome_cais_slopes.csv')

```
