---
title: "PIC"
author: "Hanon McShea"
date: "2024-02-28"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(ape)
library(MASS)
library(nlme)
library(lme4)
require(lmerTest)
library(optimx)
library(ggplot2)
library(MuMIn)
library(plotly)
library(plyr)
library(phytools)
library(scatterplot3d)
library(rgl)
library(gridExtra)
library(caroline)
```

## Read in data files
```{r data}
setwd("C:/Users/hanon/Documents/GitHub/aa_flux/CAIS/")

#read in newick tree
tree<-read.tree(file = "PhylogeneticTree_AllSpecies.nwk.txt")

#read in amino acid frequencies, both proteome or Pfam
pf_aaf<-read.csv(file = "AllPfam_AAC.csv", header = T)
pr_aaf<-read.csv(file = "VertebrateAAC_CAIS.csv", header = T)
pr_aaf<-pr_aaf[,2:24]
pr_aaf<-distinct(pr_aaf, SpeciesUID, .keep_all=T)
cais<-read.csv(file="CAIS_KLD.txt", header=T)
species<-read.csv(file="SpeciesList_11_7_2019.txt",header=T)

colnames(cais)[1] <- 'SpeciesUID'

#decide from here on out whether to use Pfam or proteome
aaf <- pf_aaf

#pf_aaf <- merge(pf_aaf, species, by.y="SpeciesUID")
#pf_aaf <- merge(pf_aaf, cais, by.y="SpeciesUID")
#pf_aaf <- subset(pf_aaf, NewickName != "Pantholops_hodgsonii") #remove antelope

#pr_aaf <- merge(pr_aaf, species, by.y="SpeciesUID")
#pr_aaf <- merge(pr_aaf, cais, by.y="SpeciesUID")
#pr_aaf <- subset(pr_aaf, NewickName != "Pantholops_hodgsonii") #remove antelope

aaf <- merge(aaf, species, by.y="SpeciesUID")
aaf <- merge(aaf, cais, by.y="SpeciesUID")
aaf <- subset(aaf, NewickName != "Pantholops_hodgsonii") #remove antelope
```

## Get data in same order as tree tips
```{r PIC prep}
#pf_namevector<-as.vector(pf_aaf$NewickName)#vector of all the vertebrate species names
#pr_namevector<-as.vector(pr_aaf$NewickName)

#pf_trimmedtree<-keep.tip(tree,pf_namevector)# keeping only the tips with vertebrate species names
#pr_trimmedtree<-keep.tip(tree,pr_namevector)

#pf_phyloorder <- as.list(pf_trimmedtree$tip.label)
#pr_phyloorder <- as.list(pr_trimmedtree$tip.label)

#pf_aafo <- arrange(pf_aaf, match(NewickName, pf_phyloorder))
#pr_aafo <- arrange(pr_aaf, match(NewickName, pr_phyloorder))

#pf_aaf
#pr_aaf

#pf_aafo
#pr_aafo

namevector<-as.vector(aaf$NewickName)#vector of all the vertebrate species names
trimmedtree<-keep.tip(tree,namevector)# keeping only the tips with vertebrate species names
phyloorder <- as.list(trimmedtree$tip.label)
aafo <- arrange(aaf, match(NewickName, phyloorder))

aafo2 <- filter(aafo, NewickName %in% phyloorder)
aaf
aafo
aafo2
```

## Phylogenetically Independent Constrasts
```{r PIC}
#Set up data into vectors for use in PIC, choosing Pfam or 
# Intialize dataframe
#AAstart <- which(colnames(pf_aafo) == 'A')
#AAend <- which(colnames(pf_aafo) == 'V')

#pictable <- data.frame(matrix(nrow = 116))

# loop model over all amino acids
#for (aa in 1:ncol(pf_aafo_pic)) {
#  vec <- nv(as.vector(aa), pf_aafo_pic$NewickName)
#  vals <- as.data.frame(pic(vec, trimmedtree))
#  colnames(vals) <- colnames(pf_aafo_pic)[aa]
#  pictable <- cbind(pictable, vals) }

#vec <- nv(as.vector(pf_aafo_pic$A), pf_aafo_pic$NewickName)
#  vals <- as.data.frame(pic(vec, trimmedtree))
#  colnames(vals) <- c("node", colnames(pf_aafo_pic)[pf_aafo_pic$A]
#  colnames(vals) <- colnames(pf_aafo_pic)[aa]
#  pictable <- cbind(pictable, vals) 


#pf_aafo_pic <- pf_aafo[,2:23]

#for(i in 1:ncol(pf_aafo_pic)) {       # for-loop over columns
#  pf_aafo_pic[ , i] <- pic(as.vector(pf_aafo_pic[ , i]), trimmedtree)
#}

#picvals <- apply(pf_aafo, 2, function(data) pic(as.vector(data), trimmedtree))


#create named vectors
Avector <- nv(as.vector(aafo$A), aafo$NewickName)
Cvector <- nv(as.vector(aafo$C), aafo$NewickName)
Dvector <- nv(as.vector(aafo$D), aafo$NewickName)
Evector <- nv(as.vector(aafo$E), aafo$NewickName)
Fvector <- nv(as.vector(aafo$"F"), aafo$NewickName)
Gvector <- nv(as.vector(aafo$G), aafo$NewickName)
Hvector <- nv(as.vector(aafo$H), aafo$NewickName)
Ivector <- nv(as.vector(aafo$I), aafo$NewickName)
Kvector <- nv(as.vector(aafo$K), aafo$NewickName)
Lvector <- nv(as.vector(aafo$L), aafo$NewickName)
Mvector <- nv(as.vector(aafo$M), aafo$NewickName)
Nvector <- nv(as.vector(aafo$N), aafo$NewickName)
Pvector <- nv(as.vector(aafo$P), aafo$NewickName)
Qvector <- nv(as.vector(aafo$Q), aafo$NewickName)
Rvector <- nv(as.vector(aafo$R), aafo$NewickName)
Svector <- nv(as.vector(aafo$S), aafo$NewickName)
Tvector <- nv(as.vector(aafo$"T"), aafo$NewickName)
Vvector <- nv(as.vector(aafo$V), aafo$NewickName)
Wvector <- nv(as.vector(aafo$W), aafo$NewickName)
Yvector <- nv(as.vector(aafo$Y), aafo$NewickName)
caisvector<-nv(as.vector(aafo$CAIS), aafo$NewickName)

vectorlist <- list(Avector, Cvector, Dvector, Evector, Fvector, Gvector, Hvector, Ivector, Kvector, Lvector, Mvector, Nvector, Pvector, Qvector, Rvector, Svector, Tvector, Vvector, Wvector, Yvector, caisvector)

#aafo_pic <- aafo[,-1]
pictable <- data.frame(matrix(nrow = 116))

for (i in vectorlist) {
  vals <- as.data.frame(pic(i, trimmedtree))
  #colnames(vals) <- gsub('vector', '', deparse(substitute(i)))
  pictable <- cbind(pictable, vals) }

pictable <- pictable[,2:22]
names(pictable) <- c("A", "C", "D", "E", "F", "G", "H", "I", "K", "L", "M", "N", "P", "Q", "R", "S", "T", "V", "W", "Y", "cais")
pictable



######################### pic comparisons, note that the -1 forces the comparisons through the origin


#Vertebrate_pic_CAIS_AAC_df <- as.data.frame(do.call(cbind, piclist))
#names(Vertebrate_pic_CAIS_AAC_df) <- c("picA", "picC", "picD", "picE", "picF", "picG", "picH", "picI", "picK", "picL", "picM", "picN", "picP", "picQ", "picR", "picS", "picT", "picV", "picW", "picY")
#Vertebrate_pic_CAIS_AAC_df


lmiPIC<-lm(pictable$K~ pictable$cais)
summary(lmiPIC)
cor.test(pictable$K, pictable$cais, method = 'pearson')
cor.test(pictable$K, pictable$cais, method = 'spearman')

# Intialize dataframe
AAstart <- which(colnames(pictable) == 'A')
AAend <- which(colnames(pictable) == 'Y')
model_estimates <- data.frame(matrix(nrow = 2))

# loop model over all amino acids
for (aa in AAstart:AAend) {

model <- lm(pictable[,aa]~pictable$cais)
model_summary <- summary(model)
model_summary <- as.data.frame(model_summary$coefficients[2,1:2])
colnames(model_summary) <- colnames(pictable)[aa]
model_estimates <- cbind(model_estimates, model_summary) }

model_estimates <- model_estimates[,-1]
rownames(model_estimates) <- c('Estimates','Std.Error')
model_estimates <- as.data.frame(t(model_estimates))

model_estimates

#save values, be sure name matches amino acid frequencies you used
write.csv(model_estimates, 'PICcorrected_Pfam_cais_slopes.csv')
#write.csv(model_estimates, 'PICcorrected_proteome_cais_slopes.csv')





#massplot2 <-ggplot(data=lmiPIC, aes(piccais,picK)) + geom_point(size=2) + labs( x = "CAIS [PIC Corrected]", y = "Adult Body Mass\n[PIC Corrected]") + stat_smooth(method = "lm", color="red")

#massplot2 <- massplot2+theme(axis.title=element_text(size=28, face ="bold"),title=element_text(size=18, face ="bold"),panel.background = element_rect(fill = "white", colour = "grey50"),panel.grid.major = element_line(colour = "grey90"),panel.grid.minor = element_line(colour = "grey90"))
#massplot2<- massplot2 + geom_text(data=lmiPIC, aes(x = -0.0005, y = -0.8, label = "Pearson's R:-0.33\n p-value:0.0065"),size=7, colour = 'black')+ theme(axis.text = element_text(size = 24)) 
#massplot2
```
